src/tls_stream.rs

```diff
@@ -798,22 +802,28 @@ impl<S> MidHandshakeTlsStream<S>
 impl<S> Write for TlsStream<S>
     where S: Read + Write
 {
+    /// In the case of a WouldBlock error, we expect another call
+    /// starting with the same input data
+    /// This is similar to the use of ACCEPT_MOVING_WRITE_BUFFER in openssl
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
         let sizes = match try!(self.initialize()) {
             Some(sizes) => sizes,
             None => return Err(io::Error::from_raw_os_error(winapi::SEC_E_CONTEXT_EXPIRED as i32)),
         };
-		
-        // We can only write if the write buffer is emptied first
-        try!(self.write_out());
-
+        
         let len = cmp::min(buf.len(), sizes.cbMaximumMessage as usize);
 
-        try!(self.encrypt(&buf[..len], &sizes));
-
+        // if we have pending output data, it must have been because a previous
+        // attempt to send this part of the data ran into an error.
+        if self.out_buf.position() == self.out_buf.get_ref().len() as u64 {
+            let left_buf = &buf[self.last_write_len..];
+            self.last_write_len += len;
+            if !left_buf.is_empty() {
+                try!(self.encrypt(left_buf, &sizes));
```

Comment:
I don't think we want to buffer up more data, but wait for what's already there to write out, and then return `Ok(self.last_write_len)`